[draft] multi-host-support [v0]

ldoktor opened the pull request:
Hello guys,

I'm done for today/this week so let's post the multi-host alpha version here. It's usable, but lacks some of the required features like the custom format and also could use some love in order to be merge-ready.

Anyway to avoid possible problems in design, let me demonstrate the usage here:

    avocado run passtest --sync-server :12345
    avocado run passtest --sync :12345

When you run booth at the same time (60s deadline) on the same machine (or provide full address and use 2 machines) should pass, the `--sync-server` contains some valuable info about synchronization.

---------------------------------------

The plan is to support support custom format (still the above will be usable, and probably better fit for CI) and the execution would be:

    avocado run simple.mht

where `simple.mht` contains:

    avocado run passtest
    avocado run passtest --remote-hostname foo

the way this should work is it creates the `--sync-server` (if not provided) and runs all lines simultaneously, extending them of `--sync`. The result would wrap booth runs in a single result. At least that's the idea.

For CI people would still probably use the first way (without the mht files).

----------------------------------------------------------------

One more concept is demonstrated in this PR, it's the `Test.avocado` class. As I mentioned on the meeting, I imagine this to be plugin-extensible, which means the `sync` plugin would only say it extends the `Test.avocado` of `barrier` method and the plugin system would link this method to the `Test.avocado` class. For now we could only use it to store "extra" features to eliminate the unnecessary clashes on user-side.

Anyway if you don't like this concept, I can remove it and move `barrier` directly into `test`.

lmr commented:
I read the patch and the description, it seems I'm lacking context to understand them.

adereis commented:
@lmr, this is an experiment in supporting multi-host testing (live migration, netperf, etc). At this point it's just an experiment and @ldoktor is proposing this architecture where avocado itself would be run multiple times in multiple machines.

Anyway, my point is that this is just an experiment and indeed, it's hard to understand. We all need more context and more discussions to mature the idea and came up with a clean and sound implementation.

ldoktor commented:
I'm sorry @lmr, I pushed this so @clebergnu and @apahim can play with that. I described and demonstrated it during our meeting. I'll post RFC on mailing list once it's readier.

PS: The `examples/tests/passtest.py` is probably the key to understand this PR.

ldoktor commented:
Btw now I noticed I forgot to include the plugin... Anyway it'll be clearer when I send the RFC. First I want to make sure avocado-vt can utilize this...

lmr commented:
@ldoktor don't worry, I wanted to make it to the scrum meeting, but had problems. I'll catch up :)

ldoktor commented:
What a wonderful day, today. Morning I added the file to "git add avocado/plugins.sync.py" and started experimenting with autotest's  multi-host (migration). Then I had problems as this is based on older code-base incompatible with current avocado-vt so I did "git checkout master", which failed so I did "git reset --hard" ... Not a good idea when you forget to "git commit" the changes you added.

So I spend afternoon with "extundelete" to find out the file is really lost (as I played with the autotest multi-host for 2 hours, which produced lots of files).

So I tried the brute force:

```python
import time

disk=open("/dev/mapper/...", "rt")

disk.seek(0, 2)
part = disk.tell() / 100
next_part = part
i = 0
disk.seek(0, 0)

pattern = "class Sync("

print time.ctime()
data1 = disk.read(20)
while data1:
        if disk.tell() > next_part:
                i += 1
                next_part += part
                print time.ctime(), i
        data2 = disk.read(2048)
        data = data1 + data2
        if pattern in data:
                print disk.tell(), data[data.find(pattern):100]
        data1 = data[-20:]

```

Which actually found 2 copies, but booth are slightly outdated. I'll post them when I adjust them to work...

Productive day, indeed...

ldoktor commented:
I love Eclipse. I didn't expected this to work, but there is one functionality I never really used (I hate it actually, because usually I need different one and I always click to this one).

I added the recovered, outdated file, `right-click->Compare With->Local history` and guess what appeared:

* 25.02.16 14.23
* 12.02.16 9:40
* 12.02.16 9:40
* ...
* 12.02.16 9:39
* 12.02.16 9:38
* ...
* 11.02.16 10:09

So next time I just have to `touch` the file and use this local history as apparently Eclipse is keeping really long history of each file I modify in it.

I'm seriously considering sending some money to Eclipse foundation...

btw the difference between the recovered and original lost file was really great:

```diff
diff --git a/avocado/plugins/sync.py b/avocado/plugins/sync.py
index 287a988..c3bd2a4 100644
--- a/avocado/plugins/sync.py
+++ b/avocado/plugins/sync.py
@@ -12,8 +12,6 @@
 # Copyright: Red Hat Inc. 2016
 # Author: Lukas Doktor <ldoktor@redhat.com>
 import time
-import atexit
-import logging
 """
 Multi-host synchronization and communication plugin
 """
@@ -24,131 +22,85 @@ import threading
 from .base import CLI
 
 
-# TODO: Create "avocado.plugins" log and use __name__ instead
-# TODO: This log must start early and re-log the content when job output dir
-#       is initialized
-LOG = logging.getLogger("avocado.test")
-
-
 class ThreadedTCPRequestHandler(SocketServer.BaseRequestHandler):
 
     barriers = {}
     lock = threading.Lock()
 
-    def _abort_barrier(self, name, host):
-        with self.lock:
-            barrier = self.barriers[name]
-            barrier["hosts"].remove(host)
-            if not len(barrier["hosts"]):
-                self.log("REMOVING BARRIER %s ABORT" % name)
-                del self.barriers[name]
-
-    def log(self, msg, level=logging.DEBUG):
-        LOG.log(level, "%s:%s: %s", self.client_address[0],
-                self.client_address[1], msg)
-
-    def _handle_barrier(self, data):
-        # Identify barrier
-        no_clients, name = data.split(':', 1)
-        no_clients = int(no_clients)
-        host = "%s:%s" % self.client_address
-        with self.lock:
-            if name not in self.barriers:
-                self.log("CREATING BARRIER %s" % name, logging.INFO)
-                self.barriers[name] = {"no_clients": no_clients,
-                                       "hosts": [host],
-                                       "handled_hosts": []}
-            elif self.barriers[name]["no_clients"] != no_clients:
-                msg = ("Barrier %s exists with different no_clients"
-                       % name)
-                self.log("ABORT:%s" % msg)
-                self.request.sendall("ABORT:%s" % msg)
-                return
-            elif len(self.barriers[name]["hosts"]) == no_clients:
-                msg = ("ABORT:Barrier %s has got enough clients: %s"
-                       % (name, self.barriers[name]))
-                self.log(msg, logging.ERROR)
-                self.request.sendall(msg)
-                return
-            else:
-                self.log("JOINING BARRIER %s" % name, logging.INFO)
-                self.barriers[name]["hosts"].append(host)
-        # Wait until barrier reached
-        barrier = self.barriers[name]
-        self.request.settimeout(1)
-        data = ""
-        while len(barrier["hosts"]) < no_clients:
-            if barrier.get("abort"):
-                self.request.settimeout(None)
-                self.log("ABORT:%s" % barrier.get("abort"))
-                self.request.sendall("ABORT:%s" % barrier.get("abort"))
-                self._abort_barrier(name, host)
-                return
-            try:
-                data += self.request.recv(1024)
-                self.log(data)
-            except socket.timeout:
-                continue
-            if data.startswith("ABORT:"):
-                barrier["abort"] = "%s: %s" % (host, data[6:])
-                self._abort_barrier(name, host)
-                return
-        # Barrier reached, make sure all clients are there
-        self.request.settimeout(None)
-        self.log("REACHED")
-        self.request.sendall("REACHED")
-        self.request.settimeout(1)
-        deadline = time.time() + 10
-        data = ""
-        while time.time() < deadline:
-            if barrier.get("abort"):
-                self.request.settimeout(None)
-                self.log("ABORT:%s" % barrier.get("abort"))
-                self.request.sendall("ABORT:%s" % barrier.get("abort"))
-                self._abort_barrier(name, host)
-                return
-            try:
-                data += self.request.recv(1024)
-                self.log(data)
-            except socket.timeout:
-                continue
-            if data == "ACK":
-                break
-        else:
-            self.request.settimeout(None)
-            msg = "%s did not responded on REACHED" % host
-            barrier["abort"] = msg
-            self.log("ABORT:%s" % msg)
-            self.request.sendall("ABORT:%s" % msg)
-            self._abort_barrier(name, host)
-            return
-        # Got ack
-        self.request.settimeout(None)
-        self.log("GOGOGO")
-        self.request.sendall("GOGOGO")
-        barrier["handled_hosts"].append(host)
-        # This host is inside barrier, clean this barrier if necessary
-        with self.lock:
-            if len(barrier["hosts"]) == len(barrier["handled_hosts"]):
-                # I'm the last, remove the barrier
-                self.log("REMOVING BARRIER %s PASS" % name, logging.INFO)
-                del self.barriers[name]
-
     def handle(self):
-        cmd, data = self.request.recv(1024).split(':', 1)
-        self.log("%s:%s" % (cmd, data))
-
-        # TODO: Add support for sending pickled objects
+        cmd, host, data = self.request.recv(1024).split(':', 2)
+        print cmd, host, data
         if cmd == "BARRIER":
-            return self._handle_barrier(data)
-        else:
-            msg = "Unknown command: %s:%s" % (cmd, data)
-            self.request.sendall("ERROR:" + msg)
-            self.log(msg, level=logging.ERROR)
+            # Identify barrier
+            no_clients, barrier = data.split(':', 1)
+            no_clients = int(no_clients)
+            with self.lock:
+                if barrier not in self.barriers:
+                    self.barriers[barrier] = {"no_clients": no_clients,
+                                              "hosts": [host],
+                                              "handled_hosts": []}
+                elif self.barriers[barrier]["no_clients"] != no_clients:
+                    msg = ("Barrier %s exists with different no_clients"
+                           % barrier)
+                    self.barriers[barrier]["abort"] = "%s: %s" % (host, msg)
+                    self.request.sendall("ABORT:%s" % msg)
+                    return
+                else:
+                    self.barriers[barrier]["hosts"].append(host)
+            # Wait until barrier reached
+            barrier = self.barriers[barrier]
+            self.request.settimeout(1)
+            data = ""
+            while len(barrier["hosts"]) < no_clients:
+                if barrier.get("abort"):
+                    self.request.settimeout(None)
+                    self.request.sendall("ABORT:%s" % barrier.get("abort"))
+                    barrier["hosts"].remove(host)
+                    return
+                try:
+                    data += self.request.recv(1024)
+                except socket.timeout:
+                    continue
+                if data.startswith("ABORT:"):
+                    barrier["abort"] = "%s: %s" % (host, data[6:])
+                    barrier["hosts"].remove(host)
+                    return
+            # Barrier reached, make sure all clients are there
+            self.request.settimeout(None)
+            self.request.sendall("REACHED")
+            self.request.settimeout(1)
+            deadline = time.time() + 10
+            data = ""
+            while time.time() < deadline:
+                if barrier.get("abort"):
+                    self.request.settimeout(None)
+                    self.request.sendall("ABORT:%s" % barrier.get("abort"))
+                    barrier["hosts"].remove(host)
+                    return
+                try:
+                    data += self.request.recv(1024)
+                except socket.timeout:
+                    continue
+                if data == "ACK":
+                    break
+            else:
+                msg = "%s did not responded on REACHED" % host
+                barrier["abort"] = msg
+                self.request.sendall("ABORT:%s" % msg)
+                barrier["hosts"].remove(host)
+                return
+            # Got ack, 
+            self.request.sendall("GOGOGO")
+            barrier["handled_hosts"].append(host)
+            # This host is inside barrier, clean this barrier if necessary
+            with self.lock:
+                if len(barrier["hosts"]) == len(barrier["handled_hosts"]):
+                    # I'm the last
+                    del barrier
 
 
 class ThreadedTCPServer(SocketServer.ThreadingMixIn, SocketServer.TCPServer):
-    pass
+    daemon_threads = True
 
 
 def parse_addr(value):
@@ -184,9 +136,6 @@ class Sync(CLI):
                            "though)")
 
     def run(self, args):
-        def exit_server(server):
-            server.shutdown()
-            server.server_close()
         if not hasattr(args, "sync_server"):
             return
         client = args.sync
@@ -201,8 +150,6 @@ class Sync(CLI):
             server_thread.setDaemon(True)
             server_thread.start()
 
-            atexit.register(exit_server, tcp_server)
-
         if client:
             if not hasattr(args, "default_multiplex_tree"):
                 raise RuntimeError("Multiplexer required by sync plugin is "

```

clebergnu commented:
Pretty cool! Glad you got your work back!

apahim commented:
ouch! I'd be in a bad mood if that was with my vim. glad you got it back :)

ldoktor commented:
Well my memory is low, so I need computer to remember everything. My `bashrc` contains:

```
export HISTFILESIZE=100000000
export HISTSIZE=100000000
```

But I didn't know about Eclipse limits. For record it's:

```
70 days
500 entries per file
10MB per file
```

I'm not sure if vim can do backups by default, I only recovered some opened files, but not the closed and removed ones.

adereis commented:
Given this is now being discussed in the context of an RFC in avocado-devel, I'm closing this PR.
